### DISCUSSION TITLE
plugin for precise audio-visual timing

### DISCUSSION THREAD

**Discussion by becky-gilbert** [jspsych/jsPsych#806] [Q&A] [answered]
Dear jsPsych-ers,

I'd like to create a plugin that would allow precise relative timing of audio and visual stimuli, i.e., show an image or text either precisely _in sync_ with the audio start time, or _during_ the audio, after some precise delay relative to the audio start time. This sort of plugin would allow the members of our lab to run cross-modal priming experiments, and I think it would be of more general use to others, so I hope that if it works well it could eventually be added to the jsPsych library.
 
I think the best way to do this is to sync the audio playback to the screen updates using requestAnimationFrame, and set the visual stimulus onset to occur after a certain number of frames relative to the audio start. Then, it would be up to the researcher to adjust their audio file so that their visual stim onset time becomes a multiple of the frame rate. For instance, if the researcher wants a word to appear exactly 90 ms after the audio starts playing, then they could add 10 ms of silence to the start of the audio file, synchronize the start of the audio to a screen update, and then show the word after 6 frames (100 ms / 16.667 = 6). I'd be interested in any feedback about this approach.

A few other things I'm interested in are whether: 
(1) there are any methods that would help to reduce the delay between when the audio playback is requested by the browser and when it actually starts playing (apart from preloading the audio file), 
(2) there's a predictable delay between the browser's request to play audio and the true audio playback time, so that this could be adjusted for, and
(3) between-system differences will introduce too much variation in the true audio onset for it to even make sense to try to do this sort of thing

Regarding 2 and 3, I have seen this recent [pre-print by Bridges et al.](https://psyarxiv.com/d6nu5/) (plus their [cool interactive table](https://psychopy.org/timing/2020/table3.html)) on stimulus and response timing in web-based software, which includes data on audio and audio-visual sync timing. So I'm wondering if people know of any other data/resources that they could point me to.

I'd be very grateful for any suggestions and/or offers of collaboration :)

Thanks,
Becky

**Comment by vijaymarupudi** [comment, marked as answer]
I'm a bit worried about the synchrony part after reading this article: https://blog.paul.cx/post/audio-video-synchronization-with-the-web-audio-api/

I found this also helpful: https://www.oreilly.com/library/view/web-audio-api/9781449332679/ch01.html#s01_8

I don't think the video API has a similar lag reporting function? It might still be possible to rely on the web audio api lag but starting the video first, getting the currentTime, and then synchronizing using the web audio api?

This sounds like a very difficult task. Thank you so much for working on it! Let me know if there's anything I can do to help (I know Javascript, but have rarely dealt with experiments that require such precise measurements!)

**Reply by jkhartshorne** [reply]
@becky-gilbert Re: #2, I think if the delays are as systematic as Bridges work (etc) suggests, then you could just literally subtract them out. The only issue is that of course this may change over time. 

@jodeleeuw and I have toyed with a few ideas over the years. I think one of them was to just embed some trials into our experiments that feed a giant database that crunches the numbers and estimates lags for different equipment. That could be done as a specific 'calibration' plugin that interested researchers could pop into their experiments -- with perhaps some benefit to them such as earlier access to the updated timings, or something.

**Reply by becky-gilbert** [reply]
Thanks very much @vijaymarupudi - those resources look really helpful. It seems to me like it's worth a shot, so I'll put something together and maybe ask for your feedback, if you don't mind.

@jkhartshorne yes I was hoping it would be possible to just factor in the average delay, which I think would work if there's not too much variability across systems. 
I love the idea of the calibration plugin and giant shared database. I'd like to do this to get data on distributions of (estimated) monitor refresh rates.

**Comment by jodeleeuw** [comment]
This primer on scheduling with the web audio API might be useful too: https://www.html5rocks.com/en/tutorials/audio/scheduling/

The application generated is here: http://webaudiodemos.appspot.com/metronome/index.html

**Reply by becky-gilbert** [reply]
Thanks!
